docs/adr/0001-pipeline-to-publish-python-packages.md

```diff
@@ -0,0 +1,36 @@
+# Pipeline to publish Python packages
+
+## Status
+
+Proposed.
+
+## Context
+
+Based on my experience over the last year using Python in a data science context, functionality typically begins life in a Jupyter notebook.  There then comes a point where some of that functionality will then be:
+
+1. Promoted into a production environment, and/or;
+1. Re-used across multiple projects within Jupyter to provide tooling in areas such as ingestion, wrangling / feature engineering or exploration / analysis.
+
+There then comes the need to "productionise" the code - for example:
+
+1. Encapsulation of code in appropriate project / class structure with a view to building core set of re-useable fucntionality;
+1. Dependency management - including managing packages that are "dev only";
+1. Automated unit tests / test coverage analysis;
+1. Linting;
+1. Publlishing as a Python package that can imported in multiple use cases : Jupter notebooks, Spark jobs and other operational processes as a first class citizen alongside 3rd party Python packages.
+
+## Decision
+
+Develop a best practice Azure DevOps pipeline (and associated documentation) that will enable a Python repository to be developed and deployed as an Azure DevOps Artefact.
```

Comment:
Given that Azure DevOps is increasingly moribund, with Microsoft having moved its attentions (and the vast majority of the staff who had been working on Azure DevOps) over to GitHub, why are we creating a new mechanism on top of Azure DevOps, instead of going straight to GitHub Actions?

Reply:
Agreed.  My thinking was to develop on Azure DevOps to get the steps pinned down and then port over to GitHub Actions thereafter given that there is perhaps more documentation / examples for Azure DevOps?  But happy either way.

Reply:
All our existing IP is build using Azure DevOps, so it'd fit in more neatly with that which might be another point in favour of using it. Mind you, our decision to use Azure DevOps for all that stuff (Corvus, Menes, Marain) was down to GitHub Actions not being ready at the point where we set all that up.

I can see arguments for either option at this point.

My main concern, though, with not going straight to GitHub Actions is that once projects have taken a dependency on this, we can't just flip over to a different build system. In the long run there would be less friction with building on GitHub from the start. Would it be possible to do a spike to see how much of an issue the relative lack of documentation is?

It's probably also a good idea to loop @JamesDawson into this conversation, because he's done most of the recent work on our build automation.

Reply:
When we last looked at ADO vs GH we looked at the GH roadmap and many of the features we wanted around releases / environments weren't scheduled until H2 2021, hence why the hybrid model. I'm not sure if that roadmap has changed at all, or whether we can come up with some other form of release / environment management?

Reply:
On the flip side - we generally don't have control over what customers choose, so our processes need to be "hostable" in either...

Reply:
I think we should be making the core of our automation more portable so that we can more easily run it on different CI/CD platforms.

Have some lower common denominator tooling that we can run anywhere to do the heavy lifting (e.g. PowerShell Core, though this might be less familiar in the broader Python eco-system) and then treat the CI/CD platform almost as the ‘chrome’ around that process.

Reply:
> 
> 
> When we last looked at ADO vs GH we looked at the GH roadmap and many of the features we wanted around releases / environments weren't scheduled until H2 2021, hence why the hybrid model. I'm not sure if that roadmap has changed at all, or whether we can come up with some other form of release / environment management?

With some recent [new GH features](https://githubuniverse.com/Continuous-delivery-with-GitHub-Actions/), we should be able to form a reasonable release management story now - so we could use this scenario to flesh-out what a "hostable" solution looks like targetting ADO and GH?

Reply:
Just discovered [this post on GitHub](https://docs.github.com/en/free-pro-team@latest/actions/guides/building-and-testing-python) which provides quite a comprehensive set of steps.  The main gap is the step to publish the package into a private package registry rather than PyPi.